composer/profiler/performance_analysis_trace_handler.py

```diff
@@ -0,0 +1,114 @@
+        training loop which may change dataloader bottlenecked status.
+        """
+        del state, logger
+        self.dataloader_wait_times = {}
+        self.batch_wait_times = {}
+
+    def epoch_start(self, state: State, logger: Logger) -> None:
+        self.reset_wait_times(state, logger)
+
+    def _record_wait_time(
+        self,
+        is_start: bool,
+        wait_times: Dict[int, Any],
+        wait_key: int,
+        wall_clock_time_ns: int,
+    ) -> None:
+        if is_start:
+            self.last_start_time = wall_clock_time_ns
+        elif self.last_start_time:
+            latency = wall_clock_time_ns - self.last_start_time
+            wait_times[wait_key] = latency
+
+    def process_duration_event(
+        self,
+        name: str,
+        categories: Union[List[str], Tuple[str, ...]],
+        is_start: bool,
+        timestamp: Timestamp,
+        wall_clock_time_ns: int,
+        action: ProfilerAction,
+    ) -> None:
+        del name
+        if 'dataloader' in categories:
+            self._record_wait_time(is_start, self.dataloader_wait_times, timestamp.batch.value, wall_clock_time_ns)
+        # Check for both [before/after]_train_batch and PerformanceAnalyzerTraceHandler so it only
+        # fires once on [before/after]_train_batch
+        elif 'before_train_batch' in categories and 'PerformanceAnalyzerTraceHandler' in categories:
+            self._record_wait_time(True, self.batch_wait_times, timestamp.batch.value, wall_clock_time_ns)
+        elif 'after_train_batch' in categories and 'PerformanceAnalyzerTraceHandler' in categories:
+            self._record_wait_time(False, self.batch_wait_times, timestamp.batch.value, wall_clock_time_ns)
+            # Only check for bottleneck on end of batch
+            if not is_start and action == ProfilerAction.ACTIVE_AND_SAVE and self.is_dataloader_bottlenecked():
+                dataloader_time = self._average_dataloader_wait_time()
+                batch_time = self._average_batch_wait_time()
+                # Log percentage overhead and additional time in milliseconds
+                percentage = format(dataloader_time / batch_time * 100, '.2f')
+                additional_latency = format(dataloader_time / 1e6, '.3f')
+                warnings.warn(f'The current training run is dataloader bottlenecked, adding {additional_latency}ms and '
+                              f'increasing batch time by {percentage}%. This warning may be spurious or noisy if the '
+                              '`active` period of the profiler schedule is too small.')
+
+    def _average_batch_wait_time(self) -> float:
+        return self._pruned_mean(list(self.batch_wait_times.values()))
+
+    def _average_dataloader_wait_time(self) -> float:
+        return self._pruned_mean(list(self.dataloader_wait_times.values()))
+
+    def _pruned_mean(self, data: List[float], mstdev_threshold=2.0) -> float:
```

Comment:
```suggestion
    def _pruned_mean(self, data: List[float], mstdev_threshold: float = 2.0) -> float:
```

Reply:
This math seems fine though curious about the rational to use this formula to exclude outliers? Is this the standard way of doing it?

Reply:
https://en.wikipedia.org/wiki/Median_absolute_deviation -- this is one way I've seen using stdev style filtering to be more robust with outliers. median stdevs are convergent in probability to stdev, but slightly more robust. Not an expert on this (recalling things I heard somewhere in a stat class)